File: javascript/src/amen/webhooks/verify.js
/** Signature verification over `timestamp + "." + rawBody` (docs/04-webhooks.md).
 * Verified against real Amen sandbox deliveries: HMAC-SHA256, hex, `sha256=` prefixed. */
import { createHmac, timingSafeEqual } from "node:crypto";
export const SIGNATURE_HEADER = "x-webhook-signature";
export const TIMESTAMP_HEADER = "x-webhook-timestamp";
export const EVENT_HEADER = "x-webhook-event";
export const ALGORITHM = "sha256";
export const computeSignature = (secret, timestamp, rawBody, algorithm = ALGORITHM) =>
  createHmac(algorithm, secret).update(String(timestamp) + ".").update(rawBody).digest("hex");
export function verifySignature(secret, timestamp, rawBody, received, algorithm = ALGORITHM) {
  if (!received || timestamp == null || timestamp === "") return false;
  const given = (received.startsWith("sha256=") ? received.slice("sha256=".length) : received).trim().toLowerCase();
  const expected = computeSignature(secret, timestamp, rawBody, algorithm);
  return given.length === expected.length && timingSafeEqual(Buffer.from(given), Buffer.from(expected));
}

File: python/src/amen/webhooks/verify.py
"""Webhook signature verification.

Verified against real Amen sandbox deliveries (2026-08-30). Amen signs the
**timestamp and the raw body together**, Stripe-style:

    signed_message = f"{X-Webhook-Timestamp}.{raw_body}"
    X-Webhook-Signature: sha256=HMAC_SHA256(secret, signed_message)

Always feed the RAW bytes and the timestamp header straight from the request —
re-serialising the JSON changes the bytes and breaks the signature.
"""
from __future__ import annotations
import hashlib, hmac

SIGNATURE_HEADER = "X-Webhook-Signature"
TIMESTAMP_HEADER = "X-Webhook-Timestamp"
EVENT_HEADER = "X-Webhook-Event"
ALGORITHM = "sha256"


def compute_signature(secret: str, timestamp: str, raw_body: bytes, algorithm: str = ALGORITHM) -> str:
    message = timestamp.encode() + b"." + raw_body
    return hmac.new(secret.encode(), message, getattr(hashlib, algorithm)).hexdigest()


def verify_signature(secret: str, timestamp: str | None, raw_body: bytes, received: str | None,
                     algorithm: str = ALGORITHM) -> bool:
    """Constant-time comparison. `received` may be prefixed like 'sha256=<hex>'."""
    if not received or not timestamp:
        return False
    received = received.split("=", 1)[1] if "=" in received else received
    expected = compute_signature(secret, timestamp, raw_body, algorithm)
    return hmac.compare_digest(expected, received.strip().lower())
